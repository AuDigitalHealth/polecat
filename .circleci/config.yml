defaults: &defaults
  working_directory: &default_workdir ~/repo
  docker:
    - image: circleci/node:8.11.3

version: 2
jobs:
  build:
    <<: *defaults
    steps:
      - checkout
      - restore_cache:
          keys:
            - v1-dependencies-{{ checksum "package.json" }}
            - v1-dependencies-
      - run: yarn install
      - save_cache:
          paths:
            - node_modules
          key: v1-dependencies-{{ checksum "package.json" }}
      - run: yarn lint
      - run: yarn test-coverage
      - run: yarn codecov
      - run: yarn build
      - persist_to_workspace:
          root: .
          paths:
            - build
  deploy:
    <<: *defaults
    steps:
      - checkout
      - attach_workspace:
          at: *default_workdir
      - add_ssh_keys
      - setup_remote_docker
      - run: yarn dockerize
      - run: yarn push-docker
      - run: ssh-keyscan $SERVER_HOSTNAME >> ~/.ssh/known_hosts
      - run: yarn deploy
      - run: yarn record-release

workflows:
  version: 2
  build_and_deploy:
    jobs:
      - build
      - deploy:
          requires:
            - build
          filters:
            branches:
              only: master
