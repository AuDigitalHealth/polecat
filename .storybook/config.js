import { configure, addDecorator } from '@storybook/react'
import StoryRouter from 'storybook-router'
import { withKnobsOptions } from '@storybook/addon-knobs'
import { withSmartKnobs } from 'storybook-addon-smart-knobs'

addDecorator(withSmartKnobs)
addDecorator(withKnobsOptions({ debounce: { wait: 350 } }))
addDecorator(StoryRouter())

const req = require.context('../src', true, /\.stories\.js$/)

function loadStories() {
  req.keys().forEach(filename => req(filename))
}

configure(loadStories, module)
