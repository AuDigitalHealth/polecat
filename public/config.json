{
  "fhirServer": "http://medserve.local/fhir"
}
