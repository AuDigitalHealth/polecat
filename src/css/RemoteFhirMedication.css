.remote-fhir-medication .loading img {
  position: absolute;
  top: 20px;
  right: 20px;
  z-index: 99;
}
