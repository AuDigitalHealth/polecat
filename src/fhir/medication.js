import difference from 'lodash.difference'
import uniq from 'lodash.uniq'

const snomedUri = 'http://snomed.info/sct'

const extensionFilterFor = key =>
  ({
    medicationResourceType: ext =>
      ext.url ===
      'http://medserve.online/fhir/StructureDefinition/medicationResourceType',
    parentMedication: ext =>
      ext.url ===
      'http://medserve.online/fhir/StructureDefinition/parentMedication',
    parentMedicationResources: ext =>
      ext.url ===
        'http://medserve.online/fhir/StructureDefinition/parentMedicationResources' &&
      ext.extension,
    brand: ext =>
      ext.url === 'http://medserve.online/fhir/StructureDefinition/brand',
  }[key])

const relationshipTypeFor = (sourceType, targetType) => {
  switch (`${sourceType}-${targetType}`) {
    // CTPP -> TPP
    // Branded package with container -> Branded package with no container
    case 'BPGC-BPG':
      return 'is-a'
    // CTPP -> TP
    // Branded package with container -> brand
    case 'BPGC-brand':
      return 'has-brand'
    // TPP -> TP
    // Branded package with no container -> brand
    case 'BPG-brand':
      return 'has-brand'
    // TPUU -> TP
    // Branded product with strengths and form -> brand
    case 'BPSF-brand':
      return 'has-brand'
    // TPP -> TPUU
    // Branded package with no container ->
    //   Branded product with strengths and form
    case 'BPG-BPSF':
      return 'has-bpsf'
    // TPUU -> MPUU
    // Branded product with strengths and form ->
    //   Unbranded product with strengths and form
    case 'BPSF-UPDSF':
      return 'is-a'
    // TPP -> MPP
    // Branded package with no container -> Unbranded package with no container
    case 'BPG-UPG':
      return 'is-a'
    // MPP -> MPUU
    // Unbranded package with no container ->
    //   Unbranded product with strengths and form
    case 'UPG-UPDSF':
      return 'has-updsf'
    // MPUU -> MP
    // Unbranded product with strengths and form ->
    //   Unbranded product with no strengths or form
    case 'UPDSF-UBD':
      return 'is-a'
    default:
      return 'unknown'
  }
}

const resourceRequirementsFor = sourceType =>
  ({
    // CTPP requires retrieval of TPUU to get MPUU and MP.
    BPGC: {
      UPDSF: 'BPSF',
      UPD: 'BPSF',
    },
    // TPP requires retrieval of TPUU to get MPUU and MP.
    BPG: {
      UPDSF: 'BPSF',
      UPD: 'BPSF',
    },
    BPSF: {},
    UPG: {
      UPD: 'UPDSF',
    },
    UPDSF: {},
    UPD: {},
  }[sourceType])

export const amtConceptTypeFor = fhirType =>
  ({
    BPGC: 'CTPP',
    BPG: 'TPP',
    BPSF: 'TPUU',
    brand: 'TP',
    UPG: 'MPP',
    UPDSF: 'MPUU',
    UPD: 'MP',
  }[fhirType])

export const getResource = parsed => {
  const code = getSctCode(parsed)
  const resourceType = parsed.extension.find(
    extensionFilterFor('medicationResourceType')
  )
  validateMedicationResourceType(resourceType)
  const type = resourceType.valueCoding.code
  return { type, ...code }
}

const getSctCode = node => {
  if (!node.code) throw new Error('Missing code element.')
  validateCoding(node.code.coding)
  const coding = node.code.coding.find(c => c.system === snomedUri)
  return { code: coding.code, display: coding.display }
}

export const getRelatedResources = (
  parsed,
  sourceConcept,
  sourceType,
  requiredResourceTypes,
  initialRelatedResources = { concepts: [], relationships: [] }
) => {
  console.log('getRelatedResources', {
    parsed,
    sourceConcept,
    sourceType,
    requiredResourceTypes,
    initialRelatedResources,
  })
  // Call the generator function and merge together the results into a single
  // object with all concepts and all relationships.
  const result = [
    ...generateRelatedResources(parsed, sourceConcept, sourceType),
  ].reduce((merged, item, initialRelatedResources) => {
    item.concepts = merged.concepts.concat(item.concepts)
    item.relationships = merged.relationships.concat(item.relationships)
    return item
  })
  // Calculate the resources that still need to be parsed in order to get all of
  // the related concept types that are required.
  result.resourcesStillNeeded = uniq(
    difference(requiredResourceTypes, result.concepts.map(obj => obj.type))
      .reduce(
        (acc, item) => acc.concat([resourceRequirementsFor(sourceType)[item]]),
        []
      )
      .filter(item => item)
  )
  return result
}

function * generateRelatedResources(node, sourceConcept, sourceType) {
  console.log('generateRelatedResources', { node, sourceConcept, sourceType })
  // Handle root of Medication resource being passed in.
  if (node.resourceType === 'Medication' && node.extension) {
    yield * generateRelatedResources(node.extension, sourceConcept, sourceType)
  } else {
    // Get parent medication.
    const parentMedication = getParentMedication(
      node,
      sourceConcept,
      sourceType
    )
    yield parentMedication
    const parentConcept = parentMedication.concepts[0]
    const newSourceConcept = parentConcept ? parentConcept.code : sourceConcept
    const newSourceType = parentConcept ? parentConcept.type : sourceType
    // Get brand.
    yield getBrand(node, sourceConcept, sourceType)
    // Recurse into each "parents" array to extract the data from each level
    // of the hierarchy.
    for (const parents of node.filter(
      extensionFilterFor('parentMedicationResources')
    )) {
      yield * generateRelatedResources(
        parents.extension,
        newSourceConcept,
        newSourceType
      )
    }
  }
}

const getParentMedication = (node, sourceConcept, sourceType) => {
  console.log('getParentMedication', { node, sourceConcept, sourceType })
  try {
    // Get the extension which provides the parent code itself.
    const parentMedication = node.find(extensionFilterFor('parentMedication'))
    validateParentMedication(parentMedication)
    // Get the extension which describes the resource type of the parent code.
    const resourceType = node.find(extensionFilterFor('medicationResourceType'))
    validateMedicationResourceType(resourceType)
    const reference = parentMedication.valueReference.reference
    // Convert the reference (Medication/[code]) to a code.
    const code = reference.split('/').slice(-1)[0]
    const display = parentMedication.valueReference.display
    const type = resourceType.valueCoding.code
    // Yield a structure with an array of concepts, and an array of
    // relationships. This will be merged with data found elsewhere in the
    // resource, and in other resources, later on.
    return {
      concepts: [{ type, code, display }],
      relationships: [
        {
          source: sourceConcept,
          target: code,
          type: relationshipTypeFor(sourceType, type),
        },
      ],
    }
  } catch (error) {
    // If `parentMedication` or `resourceType` are missing or malformed, don't
    // add anything to the result.
    return { concepts: [], relationships: [] }
  }
}

const getBrand = (node, sourceConcept, sourceType) => {
  console.log('getBrand', { node, sourceConcept, sourceType })
  try {
    // Get the extension which provides the brand information.
    const brand = node.find(extensionFilterFor('brand'))
    validateBrand(brand)
    const coding = brand.valueCodeableConcept.coding.find(
      c => c.system === snomedUri
    )
    validateCoding(brand.valueCodeableConcept.coding)
    const code = coding.code
    const display = coding.display
    return {
      concepts: [{ type: 'brand', code, display }],
      relationships: [
        {
          source: sourceConcept,
          target: code,
          type: relationshipTypeFor(sourceType, 'brand'),
        },
      ],
    }
  } catch (error) {
    return { concepts: [], relationships: [] }
  }
}

const validateParentMedication = parentMedication => {
  if (!parentMedication) throw new Error('Missing parentMedication value.')
  if (!parentMedication.valueReference) {
    throw new Error('Missing parentMedication.valueReference.')
  }
  if (!parentMedication.valueReference.reference) {
    throw new Error('Missing parentMedication.valueReference.reference')
  }
  if (!parentMedication.valueReference.display) {
    throw new Error('Missing parentMedication.valueReference.display')
  }
}

const validateMedicationResourceType = medicationResourceType => {
  if (!medicationResourceType) {
    throw new Error('Missing medicationResourceType value.')
  }
  if (!medicationResourceType.valueCoding) {
    throw new Error('Missing medicationResourceType.valueCoding.')
  }
  if (!medicationResourceType.valueCoding.code) {
    throw new Error('Missing medicationResourceType.valueCoding.code.')
  }
}

const validateBrand = brand => {
  if (!brand) throw new Error('Missing brand value.')
  if (!brand.valueCodeableConcept) {
    throw new Error('Missing brand.valueCodeableConcept.')
  }
}

const validateCoding = coding => {
  if (!coding) throw new Error('Missing coding.')
  for (const c of coding.filter(c => c.system === snomedUri)) {
    if (!c.code) throw new Error('Missing coding.code value.')
    if (!c.display) throw new Error('Missing coding.display value.')
  }
}
