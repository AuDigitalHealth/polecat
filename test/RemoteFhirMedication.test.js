import React from 'react'
import { shallow } from 'enzyme'
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'

import { RemoteFhirMedication } from '../src/RemoteFhirMedication.js'
import FhirMedication from '../src/FhirMedication.js'
import { fhirMedicationTypeFor } from '../src/fhir/medication.js'
import config from './config.js'

import ctpp1 from './fixtures/ctpp-1065531000168104.json'
import ctpp2 from './fixtures/ctpp-19255011000036102.json'
import substance1 from './fixtures/substance-2292011000036106.json'
import childBundle from './fixtures/childBundle-21360011000036101.json'
import packageBundle from './fixtures/packageBundle-53975011000036108.json'
import containsIngredientBundle from './fixtures/ingredientBundle-1975011000036109.json'
import operationOutcome from './fixtures/operationOutcome.json'

const mock = new MockAdapter(axios)

describe('RemoteFhirMedication', () => {
  afterEach(() => mock.reset())

  it('should pass children on to FhirMedication', () => {
    const props = {
      fhirServer: config.fhirServer,
      id: '1065531000168104',
    }
    const wrapper = shallow(
      <RemoteFhirMedication {...props}>
        <div id="someChild" />
      </RemoteFhirMedication>,
    )
    // Find the child element inside the FhirMedication component.
    const child = wrapper.find(FhirMedication).find('#someChild')
    expect(child).toBeDefined()
  })

  describe('mount', () => {
    it('should provide Medication resource to FhirMedication', () => {
      const props = {
        fhirServer: config.fhirServer,
        id: '1065531000168104',
        // `resourceType` should default to "Medication".
      }
      mock
        // Stub out the HTTP request to retrieve the subject resource.
        .onGet(`${props.fhirServer}/Medication/${props.id}`)
        .replyOnce(200, ctpp1, { 'content-type': 'application/fhir+json' })
      const wrapper = shallow(
        <RemoteFhirMedication {...props}>
          <div />
        </RemoteFhirMedication>,
      )
      // Give the async code a chance to execute, then check that the subject
      // resource has been passed through to the `resource` prop of the
      // FhirMedication component.
      return new Promise(resolve => {
        setTimeout(() => {
          wrapper.update()
          const fhirMedication = wrapper.find(FhirMedication)
          expect(fhirMedication).toBeDefined()
          expect(fhirMedication.prop('resource')).toBeDefined()
          expect(fhirMedication.prop('resource')).toEqual(ctpp1)
          resolve()
        }, 50)
      })
    })

    it('should provide Substance resource to FhirMedication', () => {
      const props = {
        fhirServer: config.fhirServer,
        id: '2292011000036106',
        resourceType: 'Substance',
      }
      mock
        // Stub out the HTTP request to retrieve the subject resource.
        .onGet(`${props.fhirServer}/Substance/${props.id}`)
        .replyOnce(200, substance1, { 'content-type': 'application/fhir+json' })
      const wrapper = shallow(
        <RemoteFhirMedication {...props}>
          <div />
        </RemoteFhirMedication>,
      )
      // Give the async code a chance to execute, then check that the subject
      // resource has been passed through to the `resource` prop of the
      // FhirMedication component.
      return new Promise(resolve => {
        setTimeout(() => {
          wrapper.update()
          const fhirMedication = wrapper.find(FhirMedication)
          expect(fhirMedication).toBeDefined()
          expect(fhirMedication.prop('resource')).toBeDefined()
          expect(fhirMedication.prop('resource')).toEqual(substance1)
          resolve()
        }, 50)
      })
    })
  })

  describe('on prop update', () => {
    it('should provide Medication resource to FhirMedication', () => {
      const props = {
        fhirServer: config.fhirServer,
        id: '1065531000168104',
      }
      const nextProps = { id: '19255011000036102' }
      mock
        // Stub out the HTTP request to retrieve the subject resource.
        .onGet(`${props.fhirServer}/Medication/${props.id}`)
        .replyOnce(200, ctpp1, { 'content-type': 'application/fhir+json' })
        // Stub out a subsequent HTTP request, triggered when the props are
        // updated to point to a new Medication ID.
        .onGet(`${props.fhirServer}/Medication/${nextProps.id}`)
        .replyOnce(200, ctpp2, { 'content-type': 'application/fhir+json' })
      const wrapper = shallow(
        <RemoteFhirMedication {...props}>
          <div />
        </RemoteFhirMedication>,
      )
      // Update props to a different ID.
      wrapper.setProps(nextProps)
      // Give the async code a chance to execute, then check that the second
      // resource has been passed through to the `resource` prop of the
      // FhirMedication component.
      return new Promise(resolve => {
        setTimeout(() => {
          wrapper.update()
          const fhirMedication = wrapper.find(FhirMedication)
          expect(fhirMedication).toBeDefined()
          expect(fhirMedication.prop('resource')).toBeDefined()
          expect(fhirMedication.prop('resource')).toEqual(ctpp2)
          resolve()
        }, 50)
      })
    })
  })

  describe('onRequireRelatedResources', () => {
    it('should pass related resources to FhirMedication', () => {
      const props = {
        fhirServer: config.fhirServer,
        id: '1065531000168104',
      }
      const relatedId = '19255011000036102'
      mock
        // Stub out the HTTP request to retrieve the subject resource.
        .onGet(`${props.fhirServer}/Medication/${props.id}`)
        .replyOnce(200, ctpp1, { 'content-type': 'application/fhir+json' })
        // Stub out the HTTP request to retrieve the related resource.
        .onGet(`${props.fhirServer}/Medication/${relatedId}`)
        .replyOnce(200, ctpp2, { 'content-type': 'application/fhir+json' })
      const wrapper = shallow(
        <RemoteFhirMedication {...props}>
          <div />
        </RemoteFhirMedication>,
      )
      // Get the `onRequireRelatedResources` callback passed to the
      // FhirMedication component.
      const callback = wrapper
        .find(FhirMedication)
        .prop('onRequireRelatedResources')
      // Call the callback with an array containing the ID of the related resource.
      callback([relatedId])
      // Give the async code a chance to execute, then check that the expected
      // resource has been passed through to the `relatedResources` prop of the
      // FhirMedication component.
      return new Promise(resolve => {
        setTimeout(() => {
          wrapper.update()
          const fhirMedication = wrapper.find(FhirMedication)
          expect(fhirMedication).toBeDefined()
          expect(fhirMedication.prop('relatedResources')).toBeDefined()
          expect(fhirMedication.prop('relatedResources')).toEqual({
            // The expected prop value is an object keyed on ID, with the
            // resource itself as the value.
            [relatedId]: ctpp2,
          })
          resolve()
        }, 50)
      })
    })
  })

  describe('onRequireChildBundle', () => {
    it('should pass bundle to FhirMedication', () => {
      const props = {
        fhirServer: config.fhirServer,
        id: '1065531000168104',
      }
      const parentId = '19255011000036102',
        resourceType = 'MP'
      mock
        // Stub out the HTTP request to retrieve the subject resource.
        .onGet(`${props.fhirServer}/Medication/${props.id}`)
        .replyOnce(200, ctpp1, { 'content-type': 'application/fhir+json' })
        // Stub out the HTTP request to search for child resources.
        .onGet(
          `${
            props.fhirServer
          }/Medication?ancestor=Medication/${parentId}&medication-resource-type=UPD`,
        )
        .replyOnce(200, childBundle, {
          'content-type': 'application/fhir+json',
        })
      const wrapper = shallow(
        <RemoteFhirMedication {...props}>
          <div />
        </RemoteFhirMedication>,
      )
      // Get the `onRequireChildBundle` callback passed to the
      // FhirMedication component.
      const callback = wrapper.find(FhirMedication).prop('onRequireChildBundle')
      // Call the callback with the parent ID, and the FHIR medication type for
      // the type of child resource to be queried.
      callback(parentId, fhirMedicationTypeFor(resourceType))
      // Give the async code a chance to execute, then check that the expected
      // bundle has been passed through to the `childBundles` prop of the
      // FhirMedication component.
      return new Promise(resolve => {
        setTimeout(() => {
          wrapper.update()
          const fhirMedication = wrapper.find(FhirMedication)
          expect(fhirMedication).toBeDefined()
          expect(fhirMedication.prop('childBundles')).toBeDefined()
          expect(fhirMedication.prop('childBundles')).toEqual({
            // The expected prop value is an object keyed on FHIR medication
            // type, with the bundle as the value.
            UPD: childBundle,
          })
          resolve()
        }, 50)
      })
    })
  })

  describe('onRequirePackageBundle', () => {
    it('should pass bundle to FhirMedication', () => {
      const props = {
        fhirServer: config.fhirServer,
        id: '1065531000168104',
      }
      const subjectId = '19255011000036102',
        resourceType = 'TPP'
      mock
        // Stub out the HTTP request to retrieve the subject resource.
        .onGet(`${props.fhirServer}/Medication/${props.id}`)
        .replyOnce(200, ctpp1, { 'content-type': 'application/fhir+json' })
        // Stub out the HTTP request to search for packages.
        .onGet(
          `${
            props.fhirServer
          }/Medication?package-item=Medication/${subjectId}&medication-resource-type=BPG`,
        )
        .replyOnce(200, packageBundle, {
          'content-type': 'application/fhir+json',
        })
      const wrapper = shallow(
        <RemoteFhirMedication {...props}>
          <div />
        </RemoteFhirMedication>,
      )
      // Get the `onRequirePackageBundle` callback passed to the
      // FhirMedication component.
      const callback = wrapper
        .find(FhirMedication)
        .prop('onRequirePackageBundle')
      // Call the callback with the subject ID, and the FHIR medication type for
      // the type of package resource to be queried.
      callback(subjectId, fhirMedicationTypeFor(resourceType))
      // Give the async code a chance to execute, then check that the expected
      // bundle has been passed through to the `packageBundles` prop of the
      // FhirMedication component.
      return new Promise(resolve => {
        setTimeout(() => {
          wrapper.update()
          const fhirMedication = wrapper.find(FhirMedication)
          expect(fhirMedication).toBeDefined()
          expect(fhirMedication.prop('packageBundles')).toBeDefined()
          expect(fhirMedication.prop('packageBundles')).toEqual({
            // The expected prop value is an object keyed on FHIR medication
            // type, with the bundle as the value.
            BPG: packageBundle,
          })
          resolve()
        }, 50)
      })
    })
  })

  describe('onRequireContainsIngredientBundle', () => {
    it('should pass bundle to FhirMedication', () => {
      const props = {
        fhirServer: config.fhirServer,
        id: '2292011000036106',
      }
      const substanceId = '2292011000036106',
        resourceType = 'TPUU'
      mock
        // Stub out the HTTP request to retrieve the subject resource.
        .onGet(`${props.fhirServer}/Medication/${props.id}`)
        .replyOnce(200, substance1, {
          'content-type': 'application/fhir+json',
        })
        // Stub out the HTTP request to search for resources containing the ingredient.
        .onGet(
          `${
            props.fhirServer
          }/Medication?ingredient=Substance/${substanceId}&medication-resource-type=BPSF`,
        )
        .replyOnce(200, containsIngredientBundle, {
          'content-type': 'application/fhir+json',
        })
      const wrapper = shallow(
        <RemoteFhirMedication {...props}>
          <div />
        </RemoteFhirMedication>,
      )
      // Get the `onRequireContainsIngredientBundle` callback passed to the
      // FhirMedication component.
      const callback = wrapper
        .find(FhirMedication)
        .prop('onRequireContainsIngredientBundle')
      // Call the callback with the substance ID, and the FHIR medication type for
      // the type of resource to be queried.
      callback(substanceId, fhirMedicationTypeFor(resourceType))
      // Give the async code a chance to execute, then check that the expected
      // bundle has been passed through to the `containsIngredientBundles` prop of the
      // FhirMedication component.
      return new Promise(resolve => {
        setTimeout(() => {
          wrapper.update()
          const fhirMedication = wrapper.find(FhirMedication)
          expect(fhirMedication).toBeDefined()
          expect(fhirMedication.prop('containsIngredientBundles')).toBeDefined()
          expect(fhirMedication.prop('containsIngredientBundles')).toEqual({
            // The expected prop value is an object keyed on FHIR medication
            // type, with the bundle as the value.
            BPSF: containsIngredientBundle,
          })
          resolve()
        }, 50)
      })
    })
  })

  describe('onLoadSubjectConcept', () => {
    it('should call the upstream callback', () => {
      const upstreamCallback = jest.fn()
      const props = {
        fhirServer: config.fhirServer,
        id: '2292011000036106',
        onLoadSubjectConcept: upstreamCallback,
      }
      const subjectConcept = {
        type: 'BPG',
        coding: [
          {
            system: 'http://snomed.info/sct',
            code: '813241000168107',
            display: 'Nuromol film-coated tablet, 24',
          },
        ],
        status: 'active',
        sourceCodeSystemUri: 'http://snomed.info/sct',
        sourceCodeSystemVersion:
          'http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180131',
        focused: true,
      }
      // Mock all HTTP requests.
      mock.onAny()
      const wrapper = shallow(
        <RemoteFhirMedication {...props}>
          <div />
        </RemoteFhirMedication>,
      )
      // Get the `onLoadSubjectConcecpt` callback passed to the FhirMedication component.
      const callback = wrapper.find(FhirMedication).prop('onLoadSubjectConcept')
      // Call the callback with the subject concept.
      callback(subjectConcept)
      // Check that the upstream callback we passed in the props of
      // RemoteFhirMedication was called with the same concept.
      expect(upstreamCallback).toBeCalledWith(subjectConcept)
    })
  })

  describe('loading', () => {
    it('should call the onLoadingChange callback on success', () => {
      const onLoadingChange = jest.fn()
      const props = {
        fhirServer: config.fhirServer,
        id: '1065531000168104',
        onLoadingChange,
      }
      mock
        // Stub out the HTTP request to retrieve the subject resource.
        .onGet(`${props.fhirServer}/Medication/${props.id}`)
        .replyOnce(200, ctpp1, { 'content-type': 'application/fhir+json' })
      shallow(
        <RemoteFhirMedication {...props}>
          <div />
        </RemoteFhirMedication>,
      )
      // Give the async code a chance to execute, then check that
      // `onLoadingChange` has been called twice - with true on the first call
      // and false on the second.
      return new Promise(resolve => {
        setTimeout(() => {
          expect(onLoadingChange.mock.calls).toHaveLength(2)
          expect(onLoadingChange.mock.calls[0][0]).toBe(true)
          expect(onLoadingChange.mock.calls[1][0]).toBe(false)
          resolve()
        }, 50)
      })
    })

    it('should call the onLoadingChange callback on error', () => {
      const onLoadingChange = jest.fn()
      const props = {
        fhirServer: config.fhirServer,
        id: '1065531000168104',
        onLoadingChange,
      }
      mock
        // Stub out the HTTP request to retrieve the subject resource.
        .onGet(`${props.fhirServer}/Medication/${props.id}`)
        .replyOnce(500)
      shallow(
        <RemoteFhirMedication {...props}>
          <div />
        </RemoteFhirMedication>,
      )
      // Give the async code a chance to execute, then check that
      // `onLoadingChange` has been called twice - with true on the first call
      // and false on the second.
      return new Promise(resolve => {
        setTimeout(() => {
          expect(onLoadingChange.mock.calls).toHaveLength(2)
          expect(onLoadingChange.mock.calls[0][0]).toBe(true)
          expect(onLoadingChange.mock.calls[1][0]).toBe(false)
          resolve()
        }, 50)
      })
    })
  })

  describe('error handling', () => {
    it('should call the onError callback on 404', () => {
      const onError = jest.fn()
      const props = {
        fhirServer: config.fhirServer,
        id: '1065531000168104',
        onError,
      }
      mock
        // Stub out the HTTP request to retrieve the subject resource with a not
        // found error.
        .onGet(`${props.fhirServer}/Medication/${props.id}`)
        .replyOnce(404)
      shallow(
        <RemoteFhirMedication {...props}>
          <div />
        </RemoteFhirMedication>,
      )
      // Give the async code a chance to execute, then check that the subject
      // resource has been passed through to the `resource` prop of the
      // FhirMedication component.
      return new Promise(resolve => {
        setTimeout(() => {
          expect(onError).toHaveBeenCalled()
          expect(onError.mock.calls[0][0].message).toBeTruthy()
          resolve()
        }, 50)
      })
    })

    it('should call the onError callback on 400 OperationOutcome', () => {
      const onError = jest.fn()
      const props = {
        fhirServer: config.fhirServer,
        id: '1065531000168104',
        onError,
      }
      mock
        // Stub out the HTTP request to retrieve the subject resource with a not
        // found error.
        .onGet(`${props.fhirServer}/Medication/${props.id}`)
        .replyOnce(400, operationOutcome, {
          'content-type': 'application/fhir+json',
        })
      shallow(
        <RemoteFhirMedication {...props}>
          <div />
        </RemoteFhirMedication>,
      )
      // Give the async code a chance to execute, then check that the subject
      // resource has been passed through to the `resource` prop of the
      // FhirMedication component.
      return new Promise(resolve => {
        setTimeout(() => {
          expect(onError).toHaveBeenCalled()
          expect(onError.mock.calls[0][0].issue).toBeTruthy()
          expect(onError.mock.calls[0][0].issue).toMatchSnapshot()
          resolve()
        }, 50)
      })
    })

    it('should call the onError callback on network error', () => {
      const onError = jest.fn()
      const props = {
        fhirServer: config.fhirServer,
        id: '1065531000168104',
        onError,
      }
      mock
        // Stub out the HTTP request to retrieve the subject resource with a not
        // found error.
        .onGet(`${props.fhirServer}/Medication/${props.id}`)
        .networkError()
      shallow(
        <RemoteFhirMedication {...props}>
          <div />
        </RemoteFhirMedication>,
      )
      // Give the async code a chance to execute, then check that the subject
      // resource has been passed through to the `resource` prop of the
      // FhirMedication component.
      return new Promise(resolve => {
        setTimeout(() => {
          expect(onError).toHaveBeenCalled()
          expect(onError.mock.calls[0][0].message).toBeTruthy()
          resolve()
        }, 50)
      })
    })

    it('should call the onError callback on timeout', () => {
      const onError = jest.fn()
      const props = {
        fhirServer: config.fhirServer,
        id: '1065531000168104',
        onError,
      }
      mock
        // Stub out the HTTP request to retrieve the subject resource with a not
        // found error.
        .onGet(`${props.fhirServer}/Medication/${props.id}`)
        .timeout()
      shallow(
        <RemoteFhirMedication {...props}>
          <div />
        </RemoteFhirMedication>,
      )
      // Give the async code a chance to execute, then check that the subject
      // resource has been passed through to the `resource` prop of the
      // FhirMedication component.
      return new Promise(resolve => {
        setTimeout(() => {
          expect(onError).toHaveBeenCalled()
          expect(onError.mock.calls[0][0].message).toBeTruthy()
          resolve()
        }, 50)
      })
    })
  })
})
