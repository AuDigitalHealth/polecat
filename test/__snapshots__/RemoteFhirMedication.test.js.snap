// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoteFhirMedication error handling should call the onError callback on 400 OperationOutcome 1`] = `
Object {
  "code": "processing",
  "diagnostics": "Invalid request: The FHIR endpoint on this server does not know how to handle GET operation[Medication] with parameters [[_text, is-brafnd]]",
  "severity": "error",
}
`;
